From 87590b689adefdd6ef9ed2ee006d2f10fcea7f76 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Mon, 6 Nov 2017 13:20:56 +0200
Subject: [PATCH 2200/2241] fix: net: mvpp2x: roll back RSS configuration

Issue:
Ethtool mv_pp2x_ethtool_set_rxfh procedure didn't RSS
Configuration if wrong parameters were set

Fix:
Save original rx_indir_table and rollback RSS configuration
if RSS set failed

This patch is part of A8K+ support series.

Change-Id: Idd54e9c786a9b43125192d507cd1f3a954b6fb0c
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/45938
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
index 26aadb2..631f770 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_ethtool.c
@@ -931,6 +931,7 @@ static int mv_pp2x_ethtool_set_rxfh(struct net_device *dev, const u32 *indir,
 {
 	int i, err;
 	struct mv_pp2x_port *port = netdev_priv(dev);
+	u32 rx_indir_table_orig[MVPP22_RSS_TBL_LINE_NUM];
 
 	if (port->priv->pp2_version == PPV21)
 		return -EOPNOTSUPP;
@@ -952,12 +953,18 @@ static int mv_pp2x_ethtool_set_rxfh(struct net_device *dev, const u32 *indir,
 	if (!indir)
 		return 0;
 
-	for (i = 0; i < ARRAY_SIZE(port->priv->rx_indir_table); i++)
+	for (i = 0; i < ARRAY_SIZE(port->priv->rx_indir_table); i++) {
+		rx_indir_table_orig[i] = port->priv->rx_indir_table[i];
 		port->priv->rx_indir_table[i] = indir[i];
+	}
 
 	err =  mv_pp22_rss_rxfh_indir_set(port);
 	if (err) {
 		netdev_err(dev, "fail to change rxfh indir table");
+		/* Rollback rx_indir_table */
+		for (i = 0; i < ARRAY_SIZE(port->priv->rx_indir_table); i++)
+			port->priv->rx_indir_table[i] = rx_indir_table_orig[i];
+		mv_pp22_rss_rxfh_indir_set(port);
 		return err;
 	}
 
-- 
2.7.4

